### ISSUE TITLE
WaveformDataRequest: adding support for time range and AVAsset.

### ISSUE THREAD

**Issue by Moriquendi** [AudioKit/AudioKit#2302]
I'm implementing an interactive waveform view which user can scroll and zoom. As the waveform is zoomed, more details should be loaded.

I like what's currently offered by the `WaveformDataRequest` but it has limitations:
1. It supports AVAudioFile only. It'd be great to add support for any AVAsset.
2. There's no support for setting a time range to load. Because of that, it's not possible to implement a "zoom-in-to-reveal-more-details" feature.

Do you think those features should be offered by AudioKit?

I'm happy to help as much as I can with implementing this but it'd be great to get your guidance.
My first thoughts about API changes:

```
WaveformDataRequest

// New initializer:
public init(asset: AVAsset) { ... }

// New property:
var timeRange: CMTimeRange { get set }
```

`getData(...)` method would not use AVAudioFile. Instead, it would switch to using `AVAssetReader`. Implementation would be similar to this: [FDWaveformRenderOperation.swift](https://github.com/fulldecent/FDWaveformView/blob/master/Sources/FDWaveformView/FDWaveformRenderOperation.swift)

Looking forward to your comments :)

**Comment by aure** [comment]
Yes, sounds great. For new features we will ask that tests be written, so familiarize yourself with AudioKit tests. This should be quite straightforward in this case to generate md5s for various zoom levels.

**Comment by Moriquendi** [comment]
Cool 🙌 
I'll work on it this month and post the update here soon.

**Comment by aure** [comment]
Cool, just make a pull request when you're ready. I don't think this needs to be an open issue for now. Feel free to comment on here though.

**Comment by Moriquendi** [comment]
As an update, I did finish a basic implementation of this that was loading waveform of a specific time range. 
However, AVAssetReader doesn't seem to be a good match for this task. My implementation is 2-5x times slower than the current one. 

I guess this shouldn't came as a surprise as Apple warns about this in their docs:

```
AVAssetReader is not intended for use with real-time sources, and its performance is not guaranteed for real-time operations.
```

So, I'll try implement this with Audio Toolbox API which hopefully will be much faster. 

I found this detailed [blog post](https://medium.com/better-programming/audio-visualization-in-swift-using-metal-accelerate-part-1-390965c095d7 ) which describes a similar problem. 
I'll go this way as well 🤞
